### ISSUE TITLE
Ability to dynamically switch between Vector and VectorImage layer

### ISSUE THREAD

**Issue by viktorijasolovjova** [openlayers/openlayers#15401]
**Problem**
We want to be able to dynamically switch between Vector and VectorImage implementations for the same layer instance depending on how many vertices there are in the currently loaded feature geometries. When vertex count is rather small we want to render the features with ol/layer/Vector, because it offers nice constant re-rendering during zoom animation, however in case we are currently displaying a lot or complex geometries this might not be performant, so we want to dynamically switch to VectorImage. 

We have tried several "hacks", all of them are not perfect:
1. We had a group layer with both Vector and VectorImage inside, one of which was set to be invisible. However we also use real layer groups in our project, so we always had to make sure to treat this kind of group differently, which caused some hard to track bugs.
2. Having one instance of Vector or VectorImage in the layer list and replacing it with another one in case we need to switch. This has a downside that the layers are getting exchanged, so it can cause bugs in case some code stored a reference to the previous layer and still assumes it is in the layer list.

**Possible solution**
We see that both `Vector` and `VectorImage` classes are very minimal and the only significant difference between them is the renderer. Our current idea is to extend `VectorImage` class with our own class and override the `getRenderer()` function, so it returns `CanvasVectorLayerRenderer` or `CanvasVectorImageLayerRenderer` depending on the current vertex count.

```
class CustomVectorLayer extends VectorImageLayer {

  #isRenderedAsImage = false;
  #renderer;
  #imageRenderer;

  constructor (...) {
    ...
    super(layerOptions);
    this.#renderer = new CanvasVectorLayerRenderer(this);
    this.#imageRenderer = new CanvasVectorImageLayerRenderer(this);
    vectorLayerSource.on(VECTOR_LAYER_SOURCE_FEATURES_LOAD_END, () => {
      this.#isRenderedAsImage = shouldBeRenderedAsImageLayer(vectorLayerSource.getFeatures());
    });
   ...
  }

  getRenderer () {
    return this.#isRenderedAsImage ? this.#imageRenderer : this.#renderer;
  }

}
```

The problem with this solution is that `ol/layer/Layer` is also sometimes accessing the field `this.renderer_` directly like for example [here](https://github.com/openlayers/openlayers/blob/82bb6790453bbd359e192c1c905d2094c3858588/src/ol/layer/Layer.js#L274). So we would like to create a merge request containing changes to such places in `ol/layer/Layer` to replace the direct usage of `this.renderer_` with the call to a getter `this.getRenderer()`. 
**Question**:  Will it be ok to merge this change when we provide it? Or maybe you have a better proposal to solve the described problem?

**Comment by viktorijasolovjova** [comment]
Thank you, right, providing the render function is a much less hacky way. Although we still extended `ol/layer/VectorImage` instead of `ol/layer/Layer`, because we want to use the vector layer functionality for this layer.

**Comment by viktorijasolovjova** [comment]
Hi, the solution with a custom `render()` function which internally uses one of the renderers (CanvasVectorLayerRenderer or CanvasVectorImageLayerRenderer) worked for us until we noticed that the feature detection with `map.getFeaturesAtPixel(pixel: Pixel)` doesn't work with it. 
The problem is that in the render function we now use the right renderer, but the feature detection still calls `layer.getRenderer()` und thus can get a wrong one.

I have an alternative proposal: could we add a protected method `disposeRenderer()` to the `ol/layer/Layer` class, which would provide an official possibility for the sub-classes to reset the renderer, so it can be completely cleaned up from the layer when needed. When during next rendering the `layer.getRenderer()` is called, it will call `layer.createRenderer()`, so another renderer will be created and will also be known to the `ol/layer/Layer`.

Our code would then look like this:
 ```
class CustomVectorLayer extends VectorImageLayer {

  #isRenderedAsImage = false;

  constructor (...) {
    ...
    super(layerOptions);
    vectorLayerSource.on(VECTOR_LAYER_SOURCE_FEATURES_LOAD_END, () => {
      this.#isRenderedAsImage = shouldBeRenderedAsImageLayer(vectorLayerSource.getFeatures());
      if (<rendering conditions changed>) {
        this.disposeRenderer();
      }
    });
   ...
  }

  createRenderer () {
    return this.#isRenderedAsImage ? new CanvasVectorLayerRenderer(this) : new CanvasVectorImageLayerRenderer(this);
  }

}
```

The change to `ol/layer/Layer` in the pull request would then be: 
```
  /**
   * Disposes and cleans up the current renderer for this layer.
   * Can be used for dynamic render switch to create another one in the next getRenderer() / createRenderer() call.
   * @protected
   */
  disposeRenderer() {
    if (this.renderer_) {
      this.renderer_.dispose();
      delete this.renderer_;
    }
  }

  /**
   * Clean up.
   */
  disposeInternal() {
    this.disposeRenderer();
    this.setSource(null);
    super.disposeInternal();
  }
```
Basically it adds a protected method `disposeRenderer` and also calls it from the existing method `disposeInternal()`. Would that be an acceptable change for you?

**Comment by viktorijasolovjova** [comment]
Closing this, because the last proposal was now done within the new WebGLVectorLayer implementation, see https://github.com/openlayers/openlayers/pull/16394. 
Since v10.3.0 `ol/layer/Layer has` a method clearRenderer which can be called to dispose the renderer.
